File: go/cryptolib/aes128_test.go
package cryptolib

import (
	"math"
	"testing"

	"github.com/candiddev/shared/go/assert"
)

func TestAES128(t *testing.T) {
	key, err := NewAES128Key()
	assert.Equal(t, err, nil)
	assert.Equal(t, len(key), int(math.Ceil(float64(16)/3)*4))

	input := []byte("testing")

	output, err := key.EncryptSymmetric(input)
	assert.Equal(t, err, nil)
	assert.Equal(t, len(output.Ciphertext), int(math.Ceil((float64(len(input))+12+4+12)/3)*4)) // IV + counter + append IV
	assert.Equal(t, output.Encryption, EncryptionAES128GCM)

	out, err := key.DecryptSymmetric(output)
	assert.Equal(t, err, nil)
	assert.Equal(t, out, input)

	// Test JavaScript outputs
	key = AES128Key("lQQyBeRmGYECoYkafl+4VQ==")
	jsOutput, _ := ParseEncryptedValue(string(EncryptionAES128GCM) + ":Guqhksdu4nsd6FIrONsXnXyDl8wS78amak0uMZ49KJSly5w=")

	output, err = key.EncryptSymmetric(input)
	assert.Equal(t, err, nil)
	assert.Equal(t, len(output.Ciphertext), int(math.Ceil((float64(len(input))+12+4+12)/3)*4)) // IV + counter + append IV

	out, err = key.DecryptSymmetric(output)
	assert.Equal(t, err, nil)
	assert.Equal(t, out, input)

	out, err = key.DecryptSymmetric(jsOutput)
	assert.Equal(t, err, nil)
	assert.Equal(t, out, input)

	key = AES128Key("UMq4DaTgIjDxPMnjm1Ajyg==")
	o, err := key.DecryptSymmetric(EncryptedValue{
		Ciphertext: "qdM5dJdTphITLfsfm/qgrpuMG2pKg4LxMjcrfe7W85U=",
		Encryption: EncryptionAES128GCM,
	})
	assert.HasErr(t, err, nil)
	assert.Equal(t, string(o), "test")
}

File: web/src/lib/encryption/AES128.test.ts
import { aes128GCMDecrypt, aes128GCMEncrypt, NewAES128Key } from "./AES128";

test("AES128", async () => {
	const key = await NewAES128Key() as string;

	expect(key.length)
		.toBe(Math.ceil(16/3) * 4);

	const input = "testing";

	let output = await aes128GCMEncrypt(key, input) as string;
	expect(output)
		.toHaveLength(Math.ceil((input.length + 12 + 4 + 12)/3) * 4); // IV + counter + append IV


	output = await aes128GCMDecrypt(key, output) as string;
	expect(output)
		.toBe(input);

	// Test Go outputs
	const goKey = "9/dd0iJw0TWcg9szIWefsA==";
	const goOutput = "tyOOaSfO6K3bx6lEVikbX7fAnIok3bu1Y9HXNPUZf3kLo3I=";

	output = await aes128GCMEncrypt(goKey, input) as string;
	expect(output)
		.toHaveLength(Math.ceil((input.length + 12 + 4 + 12)/3) * 4); // IV + counter + append IV

	output = await aes128GCMDecrypt(goKey, output) as string;
	expect(output)
		.toBe(input);

	output = await aes128GCMDecrypt(goKey, goOutput) as string;
	expect(output)
		.toBe(input);
});
